src/chapters/angular-lsn2/event-functions.rst

```diff
@@ -0,0 +1,212 @@
+
+   .. sourcecode:: TypeScript
+      :linenos:
+
+      export class MovieListComponent implements OnInit {
+         movies = ['Toy Story', 'The Shining', 'Sleepless in Seattle', 'The Martian'];
+
+         addMovie (newTitle: string) {
+            this.movies.push(newTitle);
+            return;
+         }
+
+         constructor() { }
+
+         ngOnInit() {
+         }
+      }
+
+   The keyword ``this`` is required.
+
+Save the changes and then refresh the page. Enter a new title to verify that it
+appears in the movie list. Your page should look something like:
+
+.. figure:: ./figures/new-movie-added.png
+   :alt: Updated movie list.
+
+Tidying Up the Display
+-----------------------
+
+Notice that after adding a new movie to the list, the text remains in the input
+box. If we click "Add" multiple times in a row, we would see something like:
+
+.. figure:: ./figures/repeated-movie.png
+   :alt: Same movie added multiple times.
+
+Let's modify the code to try to prevent this from happening.
+
+Clear the Input Box
+^^^^^^^^^^^^^^^^^^^^
+
+5. After the user submits a new title, we can clear the input box by setting its
+   value to be the empty string (``""``). Open ``movie-list.component.html``
+   and modify the input statement as follows:
+
+   .. sourcecode:: html+ng2
+
+      <input #newMovie (keyup.enter)='addMovie(newMovie.value); newMovie.value = ""' type='text' placeholder="Enter Movie Title Here"/>
+
+   When ``keyup.enter`` occurs, the code calls ``addMovie``. Once control
+   returns from the function, ``newMovie.value`` is set equal to ``""``, which
+   clears any text from the input box.
+
+#. Since the user can also click the "Add" button to submit a title, we need to
+   modify the ``<button>`` element as well:
+
+   .. sourcecode:: html+ng2
+
+      <button (click)='addMovie(newMovie.value); newMovie.value = ""'>Add</button>
+
+   Now ``newMovie.value`` is set equal to ``""``, when "Enter" or "Add" are used
+   to submit data.
+
+.. admonition:: Try It
+
+   Refresh the page and verify that the input box gets cleared after each new
+   title.
+
+Check for Duplicates
+^^^^^^^^^^^^^^^^^^^^^
+
+Even though we clear the input box, there is nothing to prevent the user from
+entering the same movie multiple times. While some fans may want to watch a
+film twenty times in a row, let's have our code prevent repeats.
+
+Recall that the :ref:`includes method <includes-examples>` checks if an array
+contains a particular element. The method gives us several ways to check for a
+repeated title. One possibility is:
+
+.. sourcecode:: TypeScript
+   :linenos:
+
+   addMovie (newTitle: string) {
+      if(!this.movies.includes(newMovie)){
+         this.movies.push(newMovie);
+      }
+      return;
+   }
+
+If the ``movies`` array already contains ``newMovie``, then the ``includes``
+method returns ``true``. The NOT operator (``!``) flips the result to
+``false``, and line 3 is skipped.
+
+.. admonition:: Note
```

Comment:
It's nice you are showing extra ways of doing this, but in my opinion this note is unnecessary.

Reply:
I hear you.

Updated.